tools/generator/dfg/sam/sam_groups.py

```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2016, Niklas Hauser
+# All rights reserved.
+
+avr_groups = \
+[
+    # SAMD devices
+    {
+        'family': ['D'],
+        'name': ['21'],
+    },{
+        'family': ['D'],
+        'name': ['51'],
+    },
+]
```

Comment:
These groups are passed to the merger, which will perform the actual device tree merging process. This is a bit try and error to see what looks good, we don't want too much noise in the resulting device files.
See also: https://blog.salkinium.com/modm-devices/#data-pipeline.

Reply:
right now it is not merging, but I am getting XML files.
what are the groupings based on?
It seems like SAM might need a `sam_peripherals.py` similar to STM32. From my experience with D21, L21, and D51 the SERCOM driver is compatible for all three except the clocks get set up slightly differently on L21/D51 and L21/D51 have a low power mode. Right now that maps to a single preprocessor #if  in the init function and a low power mode function that does nothing for D21. that might be two different SERCOM implementations?

Reply:
Maybe try lower-case `d` for family?
The groupings are established manually (the resulting merged tree “looks structured” -> good group), I thought about an automatic heuristic approach, but trial and error approach is so much easier.

Yes, you will want to add a `sam_peripherals.py`. Currently it is the convention to prefix the driver type with the `{platform}-type` or just `{platform}` if its unique, so that the lbuild modules do not need to check both the identifier and the driver type.

Reply:
Regarding the SERCOM, if the difference is so minor, make them the same type, but maybe add `feature` elements, which can distinguish smaller differences between peripherals. However, we usually only find these differences during the HAL implementation, and then encode this in the HAL until it becomes messy and then move it to the device files. So don’t sweat the details for now.